### PR TITLE
SAAS-30 Execute checks

### PR DESCRIPTION
**Pull request by AlekSi**

### CODE REVIEW

File: services/alertmanager/alert.go [hunk 1]
```diff
@@ -27,6 +27,7 @@ import (
 )
 
 // Severity defines alert severity.
+// TODO Move to saas package: https://jira.percona.com/browse/SAAS-84
 type Severity string
```
Comment (AlekSi) [inline services/alertmanager/alert.go:31]: @artemgavrilov ^

File: services/alertmanager/registry.go [hunk 2]
```diff
@@ -76,14 +70,14 @@ func (r *registry) RemovePrefix(prefix string) {
...
-func (r *registry) Collect() ammodels.PostableAlerts {
+func (r *Registry) Collect() ammodels.PostableAlerts {
 	r.rw.RLock()
 	defer r.rw.RUnlock()
 
 	var res ammodels.PostableAlerts
 	now := time.Now()
 	for id, t := range r.times {
-		if t.After(now) {
+		if t.Before(now) {
```
Comment (artemgavrilov) [inline services/alertmanager/registry.go:80]: Love it)

Comment (AlekSi) [reply services/alertmanager/registry.go:80]: LOL. I remember agonising on that line, changing it several times.

@atymchuk that explains disappearing alerts

Comment (atymchuk) [reply services/alertmanager/registry.go:80]: Cool, then the fix is not insanely difficult :)

File: services/alertmanager/alert.go [hunk 3]
```diff
@@ -60,78 +47,9 @@ func (ap *AlertParams) validate() error {
 		return errors.New("empty Description")
 	}
 
-	if ap.Severity == "" {
-		ap.Severity = Info
+	if ap.Severity < check.Emergency || ap.Severity > check.Debug {
```
Comment (AlekSi) [inline services/alertmanager/alert.go:50]: I think we should verify that only when we check results, so we don't have this check in two places.

File: services/checks/checks.go [hunk 4]
```diff
@@ -110,12 +115,22 @@ func (s *Service) Run(ctx context.Context) {
 	defer ticker.Stop()
 
 	for {
-		// FIXME do that only if STT is enabled in settings https://jira.percona.com/browse/SAAS-30
-		if true {
+		var sttEnabled bool
```
Comment (AlekSi) [inline services/checks/checks.go:123]: Talked with Roma – yes, we should just delay the first check for one minute.

File: services/checks/checks.go [hunk 5]
```diff
@@ -200,38 +215,32 @@ func (s *Service) executeMySQLChecks(ctx context.Context, checks []check.Check)
...
 	for _, target := range targets {
-		r, err := models.CreateActionResult(s.db.Querier, target.agentID)
+		r, err := models.CreateActionResult(s.db.Querier, target.agent.AgentID)
 		if err != nil {
-			s.l.Errorf("Failed to prepare action result for agent %s: %s.", target.agentID, err)
+			s.l.Errorf("Failed to prepare action result for agent %s: %s.", target.agent.AgentID, err)
 			continue
 		}
 
 		for _, c := range checks {
```
Comment (AlekSi) [inline services/checks/checks.go:229]: What I don't particularly like in that code that there we have two sets of checks: one is passed as an explicit parameter, the other is available as implicit parameters s.checks. I think we should get rid of `s.checks` completely and always values as parameters.

File: services/checks/checks.go [hunk 6]
```diff
@@ -316,26 +319,82 @@ func (s *Service) executeMongoDBChecks(ctx context.Context, checks []check.Check
...
+}
+
 // target contains required info about check target
 type target struct {
-	agentID   string
-	serviceID string
-	dsn       string
+	agent   *models.Agent
+	service *models.Service
+	node    *models.Node
```
Comment (AlekSi) [inline services/checks/checks.go:401]: I think both `agent` and `node` are not actually used – they were added for demo.

File: services/alertmanager/alert.go (hunk 3 above)
Comment (artemgavrilov) [reply services/alertmanager/alert.go:50]: In fact that function is unused at all.

File: services/checks/checks.go (hunk 5 above)
Comment (artemgavrilov) [reply services/checks/checks.go:250]: The idea with `s.checks` is that when for some reason we can't download new checks form server we still able to run previously downloaded stuff.

File: services/checks/checks.go (hunk 4 above)
Comment (artemgavrilov) [reply services/checks/checks.go:138]: Timeout added

File: services/checks/checks.go (hunk 6 above)
Comment (artemgavrilov) [reply services/checks/checks.go]: All that stuff needed to create labels set. Refactored, now targen contains labels set instead of node/service/agent.

**Comment by artemgavrilov** [comment]
Resolving conflicts

File: services/alertmanager/alert.go (hunk 3 above)
Comment (AlekSi) [reply services/alertmanager/alert.go]: The whole file is not used: 59fe895223e541a6e82a4e9d9aef469a550bd815

File: services/checks/checks.go [hunk 7]
```diff
@@ -166,55 +204,69 @@ func (s *Service) waitForResult(ctx context.Context, resultID string) (*models.A
...
-		return res, nil
+		return out, nil
 	}
 }
 
 // executeChecks runs all available checks for all reachable services.
 func (s *Service) executeChecks(ctx context.Context) {
+	s.l.Info("Executing checks...")
+
 	mySQLChecks, postgreSQLChecks, mongoDBChecks := s.groupChecksByDB(s.checks)
```
Comment (AlekSi) [inline services/checks/checks.go:220]: That's why state is bad :) We are not holding a lock there.
Will fix.

Comment (artemgavrilov) [reply services/checks/checks.go:220]: oops, yeah, checks should be accessed via `s.Checks` method

**Pull request merged**